packages/SwingSet/src/vats/vattp/vat-vattp.js

```
@@ -29,21 +34,60 @@ function makeCounter(render, initialValue = 7) {
-  // { outbound: { highestRemoved, highestAdded },
-  //   inbound: { highestDelivered, receiver } }
-
-  function getRemote(name) {
-    if (!remotes.has(name)) {
-      remotes.set(name, {
-        outbound: { highestRemoved: 0, highestAdded: 0 },
-        inbound: { highestDelivered: 0, receiver: null },
-      });
+
+  const mailboxKindHandle = provideKindHandle(baggage, 'Mailbox');
+  const initMailboxState = () => ({
+    outboundHighestAdded: 0,
+    outboundHighestRemoved: 0,
+    inboundHighestDelivered: 0,
+    inboundReceiver: null,
+  });
+  const makeMailbox = defineDurableKindMulti(
+    mailboxKindHandle,
+    initMailboxState,
+    {
+      outbound: {
+        getHighestAdded: ({ state }) => state.outboundHighestAdded,
+        setHighestAdded: ({ state }, highestAdded) => {
+          assert(highestAdded >= state.outboundHighestAdded);
+          state.outboundHighestAdded = highestAdded;
+        },
+        getHighestRemoved: ({ state }) => state.outboundHighestRemoved,
+        setHighestRemoved: ({ state }, highestRemoved) => {
+          assert(highestRemoved >= state.outboundHighestRemoved);
+          state.outboundHighestRemoved = highestRemoved;
+        },
+      },
+      inbound: {
+        getReceiver: ({ state }) => state.inboundReceiver,
+        initReceiver: ({ state }, receiver) => {
+          assert.equal(state.inboundReceiver, null);
+          state.inboundReceiver = receiver;
+        },
+        getHighestDelivered: ({ state }) => state.inboundHighestDelivered,
+        setHighestDelivered: ({ state }, highestDelivered) => {
+          assert(highestDelivered >= state.inboundHighestDelivered);
+          state.inboundHighestDelivered = highestDelivered;
+        },
+      },
+    },
+  );
+
+  if (!baggage.has('mailboxes')) {
+    baggage.init(
+      'mailboxes',
+      makeScalarBigMapStore('mailboxes', { durable: true }),
+    );
+  }
+  const mailboxes = baggage.get('mailboxes');
+  function provideMailbox(name) {
```

Instead of `provideMailbox`, our `addRemote` should use `mailboxes.init(name, makeMailbox())`, and our `deliverInboundMessages/Ack` should use `mailboxes.get`. Every remote name should either be added or not, and `deliverInboundMessages('unknown remote name')` should throw an error rather than silently create a new Mailbox.

---

Well, except that maybe we won't use those names, see below

---

https://github.com/Agoric/agoric-sdk/issues/5824